Move `xcdatamodeld` into WordPressData

mokagio opened the pull request:
Part of #24165 

---

> [!NOTE]  
> As I try to move more model files, I keep pulling at the intricate dependencies that grew in the codebase.
> I'll be leaving comments in the diff as starting point for code to extract in order to enable this PR.

<!-- blue -->
> [!NOTE]  
> Work in Progress. So far:
>
> - Moved the `xcdatamodeld` folder into the module and updated the code that referenced its URL to read it from the module
> - Run the app and saw it crash because it could not load `BloggingPrompt` (apparently the first model to load at runtime). The error was with `Fatal error: NSArray element failed to match the Swift Array Element type
. Expected BloggingPrompt got NSManagedObject`.
> - Moved `BloggingPrompt` to WordPressData
> - App still crashed with `Fatal error: NSArray element failed to match the Swift Array Element type
. Expected BloggingPrompt got NSManagedObject`
> - The only way I found to fix this was to hardcode the module name in the BloggingPrompt definition in the Core Data schema GUI
>
> I don't like this solution because it would require editing the schema for each model. Besides, the compiler (or whichever piece is involved in going from the schema to the binary) should be able to inject the correct module name without hardcoding.
>
> Leaving this up as a draft so that others can look at it.
>
> The next thing I'll try will be to move more of the components involved in making the Core Data fetch request, the `NSManagedObjectContext` in particular, and see if it gets the namespace resolution to work without hardcoding.

---

Fixes #

To test:

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

mokagio committed:
Update `.editorconfig` to indent Swift with 4 spaces outside Xcode

mokagio committed:
Extract definition for `Bundle` where DB lives

mokagio committed:
App crashes – Move `WordPress.xcdatamodeld` to WordPressData

mokagio committed:
App crashes – Move `BloggingPrompt` to WordPressData

mokagio committed:
Hardcode module name in `BloggingPrompt` model – Fixes the crash?!

dangermattic commented:
<!--
  0 Errors
  3 Warnings: <code>Modules/Package.swift</c...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio commented on Modules/Sources/WordPressData/WordPress.xcdatamodeld/WordPress 154.xcdatamodel/contents:
```diff
@@ -215,7 +215,7 @@
         <relationship name="blog" optional="YES" maxCount="1" deletionRule="Nullify" destinationEntity="Blog" inverseName="authors" inverseEntity="Blog" syncable="YES"/>
     </entity>
-    <entity name="BloggingPrompt" representedClassName=".BloggingPrompt" syncable="YES">
+    <entity name="BloggingPrompt" representedClassName="WordPressData.BloggingPrompt" syncable="YES">
```
Before 

![image](https://github.com/user-attachments/assets/7a66492c-a117-46ba-b2ea-936d111bdd5a)

After 

![image](https://github.com/user-attachments/assets/910ff521-6ff0-4750-855d-6910aec2be12)

mokagio commented on Modules/Sources/WordPressData/WordPressData.swift:
```diff
@@ -1 +1,29 @@
 @_exported import WordPressDataObjC
+
+public let modelURL: URL = {
+    guard let url = Bundle.module.url(forResource: "WordPress", withExtension: "momd") else {
+        fatalError("Cannot find model file.")
+    }
+    return url
+}()
+
+public func urlForModel(name: String, in directory: String?) -> URL? {
+    let bundle = Bundle(for: TemporaryDummyClassToPickUpModule.self)
+    var url = bundle.url(forResource: name, withExtension: "mom", subdirectory: directory)
+
+    if url != nil {
+        return url
+    }
+
+    let momdPaths = bundle.paths(forResourcesOfType: "momd", inDirectory: directory)
+    momdPaths.forEach { (path) in
+        if url != nil {
+            return
+        }
+        url = bundle.url(forResource: name, withExtension: "mom", subdirectory: URL(fileURLWithPath: path).lastPathComponent)
+    }
+
+    return url
+}
```
These free functions should go into an object at some point.

It was just simpler to chuck them in here at this stage while I figure things out.

Also, I'd like to think there will be no need to have them `public` once all the types involved in setting up the Core Data stack have been moved.

kean commented:
Have you tried overriding an Objective-C class name for the managed objects written in Swift?

```
[@objc](BloggingPrompt)
class BloggingPrompt: NSManagedObject {
```

mokagio commented:
> Have you tried overriding an Objective-C class name for the managed objects written in Swift?
> 
> ```
> [@objc](BloggingPrompt)
> class BloggingPrompt: NSManagedObject {
> ```

@kean yes I did try that. I removed it from the code pushed here because it did not make any difference

mokagio committed:
Move `NSManagedObject+Lookup` to WordPressData

mokagio committed:
Add CocoaLumberjack to WordPressData as a dependency

mokagio committed:
Make `AsyncOperation` and `AsyncBlockOperation` `public`

This is so the diff is neater next when moving it to WordPressShared

mokagio committed:
Move `AsyncOperation` and `AsyncBlockOperation` to WordPressShared

Turned out there needed to be move access control changes than
those done in the previous commit...

mokagio committed:
WIP – Move ContextManager and all it needs to compile to WordPressData

mokagio committed:
Fix `overrideInstance` testing property concurrency issue

Not the best approach, but fine because the property is for testing
only.

mokagio committed:
Use Swift 5 in WordPressData to avoid having to support `Sendable` ATM

We'll obviously have to deal with this at some point, but doing it while
moving the files from the app targets to the new module is not the right
time.

mokagio committed:
Swift 5 in WordPressData tests

mokagio committed:
Replace a couple of `sharedInstance()` calls to `shared`

Swift-ier + Saves making `sharedInstance()` public when migrating

mokagio committed:
Make WordPressDataObj depend on CocoaLumberJack and WordPressShared

Currently unused, but we'll use them soon when moving more of the
Objective-C models.

mokagio committed:
WIP - Move various ObjC model - BUT NOTICE THE COMMENTED OUT CODE!!!

mokagio committed:
WIP - Move more files - Build is broken

mokagio committed:
Make `entityName()` open to see if compilation error goes away...

mokagio committed:
Delete leftover versions of some of the files moved

mokagio committed:
Declare a `sharedInstance` in `CoreDataStack`

This is to decouple the Objective-C implementation from knowing about
components (i.e. `ContextManager`) in the Swift layer.

mokagio committed:
Add `import UIKit` to some files shared with notification extension

All those Swift files compile without importing UIKit because
`SFHFKeychainUtils.h` did that for them and they got access to it via
the bridging header.

Next, we'll move SFHFKeychainUtils into the modules, and UIKit will no
longer be available through it.

Side note: UIKit should not have been available via SFHFKeychainUtils in
the first place. On top of that, SFHFKeychainUtils does not need UIKit
either.

mokagio committed:
Extract SFHFKeychainUtils in a dedicated Swift package

We need it in a dedicated package because the Objective-C implementation
does not use ARC. Therefore, we need to use the `-fno-objc-arc` compiler
setting. But such a setting should be applied sparingly so instead of
blanket using it in an existing Objective-C package like
WordPressSharedObjC, we use it surgically just for this file.

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into mokagio/move-xcdatamodeld-to-wordpressdata

mokagio committed:
Remove duplicated `CoreDataStackMock` class

mokagio committed:
Hack code in `Blog.m` to attempt compilation

mokagio committed:
Move even more files — App still does not compile

mokagio commented:
Closing in favor of #24242
